Sources/StreamChat/ChatClient.swift

```diff
@@ -348,10 +348,12 @@ public class ChatClient {
         )
     }
     
-    /// Disconnects the chat client from the chat servers. No further updates from the servers
-    /// are received.
-    public func disconnect() {
-        clientUpdater.disconnect()
+    /// Stops active web-socket connection and cancels all queued API requests.
+    /// If there is a connection recovery flow running, it will be interrupted.
+    ///
+    /// - Parameter completion: The completion handler invoked when the `ChatClient` has been disconnected.
+    public func disconnect(completion: (() -> Void)? = nil) {
+        clientUpdater.disconnect { completion?() }
```

Comment:
Do we need to expose this completion to the `ChatClient`? How would this be useful for the customers?

Reply:
I don't have a concrete use-case in mind but someone might be interested when the disconnection process (or `logout` that we've discussed recently) is completed. Show loaded that the disconnection is happening/keeping the chat screen opened till chat-client is disconnected/sending a request to the backend/tracking an analytics event). It does not cost us anything to expose it so why not to have it?

Reply:
> Show spinner when the disconnection is happening/keeping the chat screen opened till chat-client

Why would customers do this? Usually, disconnection is a fire and forget action.

> It does not cost us anything to expose it so why not have it?

It will cause confusion to the customers. They will think "Should I actually wait for the disconnection to be completed?", "What if I don't wait, what is the problem?". 

IMO, I think, for now, we should not add this completion. It will add confusion to the customers, and in the future we might not even need this completion at all. So for this reasons, I would not introduce this. If in the future we encounter a valid use case, or a customer wants this, we can introduce it.

Reply:
> if we add it now, we can't remove it

But why would we limit an integrator to know when disconnection is completed? The param is `nil` by default so if they are not interested in completion they choose not to pass it

<img width="1288" alt="Screenshot 2022-05-19 at 14 51 06" src="https://user-images.githubusercontent.com/12818985/169276996-c4baa061-5d16-4edb-9753-0d8498b79148.png">

Reply:
We have optional `completion` in every async method in SDK so I wouldn't make that one an exception 🙂

Reply:
I think we can’t compare this one with the other methods you are describing. Because on the other ones we complete with an Error, so that is a valid use case to wait for it to complete and make an action if there is an error. In this case, disconnect calls usually are fire and forget. Adding a completion will only add confusion to the customers. There's no reason they should wait for a disconnect call to finish. Imagine closing a Chat UI, and waiting a little bit before it closes, this doesn't make sense.

We need to think if there is actually a valid case for this completion. If there isn't, we shouldn't add a new API that we can't remove it in the future. IMO we should try to minimise the risks here. We can always add a new API in the future, but we can't remove it. 

I have one question tho: What happens if you disconnect, and connect very fast. Is there any problem with this? Can a user connect before the disconnect finishes? 

CC @GetStream/ios-developers

Reply:
PR description:
>Fix crash happening when the chat client is connected with another user. The issue was connected to the last batch of events from the previous user posted after the new user is connected. The crash's root cause were models held by events that no longer exist after database removal.

so @nuno-vieira 
> I have one question tho: What happens if you disconnect, and connect very fast. Is there any problem with this? Can a user connect before the disconnect finishes?

it's a race condition, and can cause a crash

Reply:
@evsaev ☝️ Is this true? If so, we are not fixing anything, we are leaving it to the customer to fix it himself and we should clarify this in the changelog

Reply:
@nuno-vieira we did fix the race condition. When the chat client is reconnected with another user, we wait for disconnection to happen before establishing the new ws connection for the new user

Reply:
see [here](https://github.com/GetStream/stream-chat-swift/blob/47f6a629021eccd1c7854565342aa8187ab2671f/Sources/StreamChat/Workers/ChatClientUpdater.swift#L51)

Reply:
Yes, that is what I understood from the PR, just wanted to double-check. CC @b-onc 

Then, I vote for removing the `completion` then. I don't see any valid use case for this, it will just confuse the customers. In this case I think we should follow YAGNI principle 😄

Reply:
I've meant before the PR, sorry for confusion.

I vote for adding the completion, since YAGNI doesn't apply to us - there's no single Y (you) for an SDK. customer 1 might not need it, customer 2 might.

Reply:
Another valid point: No other SDK has this callback in the disconnect call. This is an implementation detail of our SDK and we need this internally to fix problems in our SDK. The disconnect call should be a fire and forget action, just like all the other platfroms.

Reply:
Changed ✅